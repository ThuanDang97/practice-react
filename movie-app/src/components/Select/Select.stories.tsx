import { Story } from '@storybook/react';
import Select from './Select';
import { TProps } from './Select';

export default {
  title: 'Components/Select',
  component: Select,
};

const Template: Story<TProps> = (args): JSX.Element => <Select {...args} />;

export const Basic: Story<TProps> = Template.bind({});
Basic.args = {};
