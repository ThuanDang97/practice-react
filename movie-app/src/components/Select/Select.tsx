import { memo, useState } from 'react';
import { sortingOptions } from '../../constants/MockData';

export type TProps = {
  onHandleSort: (option: string) => void;
};

const Select = (props: TProps): JSX.Element => {
  const { onHandleSort } = props;
  const [isEnable, setIsEnable] = useState(false);
  const [titleOption, setTitleOption] = useState('Popularity Descending');

  const handlerOpenSubmenu = () => {
    setIsEnable(!isEnable);
  };

  const handleSetTitle = (sortBy: string, title: string) => {
    setTitleOption(title);
    setIsEnable(!isEnable);
    onHandleSort(sortBy);
  };

  const renderSortSubMenu = (
    <div className="sub-menu" data-testid="sub-menu">
      <ul className="sub-list">
        {sortingOptions.map((item) => (
          <li
            data-testid="item"
            className="item"
            key={item.id}
            onClick={() => handleSetTitle(item.sort_by, item.title)}
          >
            {item.title}
          </li>
        ))}
      </ul>
    </div>
  );

  return (
    <>
      <span className="dropdown" onClick={handlerOpenSubmenu} data-testid="select">
        <span className="input" data-testid="titleOption">
          {titleOption}
        </span>
        <span className="select">
          <span className="icon-select"></span>
        </span>
      </span>
      {isEnable ? renderSortSubMenu : ''}
    </>
  );
};

export default memo(Select);
