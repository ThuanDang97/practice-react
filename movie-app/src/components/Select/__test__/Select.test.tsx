import { screen, render } from '@testing-library/react';
import { setup } from '../../../support/Setup';
import Select from '../Select';
import { TProps } from '../Select';

describe('Test Select component', () => {
  const defaultProps: TProps = {
    onHandleSort: jest.fn(),
  };

  test('Component Select matches DOM Snapshot', () => {
    const component = <Select {...defaultProps} />;
    const { container } = render(component);

    expect(container).toMatchSnapshot();
  });

  test('Component Select should render default correctly', () => {
    setup(<Select {...defaultProps} />);

    expect(screen.getByTestId('titleOption')).toHaveTextContent('Popularity Descending');
  });
  test('renderSortSubMenu should render after click on dropdown', async () => {
    const { user } = setup(<Select {...defaultProps} />);

    await user.click(screen.getByTestId('select'));
    expect(screen.getByTestId('sub-menu')).toBeInTheDocument();
  });

  test('Title option must change with selected option', async () => {
    const { user } = setup(<Select {...defaultProps} />);

    await user.click(screen.getByTestId('select'));
    await user.click(screen.getByText('Title (A-Z)'));
    expect(screen.getByTestId('titleOption')).toHaveTextContent('Title (A-Z)');
  });
});
