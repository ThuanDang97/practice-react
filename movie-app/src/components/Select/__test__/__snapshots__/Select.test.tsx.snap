// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Select component Component Select matches DOM Snapshot 1`] = `
<div>
  <span
    class="dropdown"
    data-testid="select"
  >
    <span
      class="input"
      data-testid="titleOption"
    >
      Popularity Descending
    </span>
    <span
      class="select"
    >
      <span
        class="icon-select"
      />
    </span>
  </span>
</div>
`;
